docs/proposals/20250513-chained-and-efficient-upgrades-for-clusters-with-managed-topologies.md

```diff
@@ -0,0 +1,420 @@
+in a Cluster API management cluster depend on external factors, e.g., by the availability of machine images for a Kubernetes version.
+
+As an alternative to explicitly setting the list of versions in a ClusterClasses, it will all also be possible to define 
+a runtime extension to be called when computing an upgrade plan; this extension could be used to return a
+dynamically computed list of Kubernetes versions that can be used.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterClass
+metadata:
+  name: quick-start-runtimesdk
+spec:
+  upgrade:
+    external:    
+      generateUpgradePlanExtension: get-upgrade-plan.foo
+```
+
+Example Request:
+
+```yaml
+apiVersion: hooks.runtime.cluster.x-k8s.io/v1alpha1
+kind: GenerateUpgradePlanRequest
+settings: <Runtime Extension settings>
+cluster:
+  apiVersion: cluster.x-k8s.io/v1beta1
+  kind: Cluster
+  metadata:
+    name: test-cluster
+    namespace: test-ns
+  spec:
+    ...
+  status:
+    ...
+fromKubernetesVersion: "v1.29.0"
+toKubernetesVersion: "v1.33.0"
+```
+
+Example Response:
+
+```yaml
+apiVersion: hooks.runtime.cluster.x-k8s.io/v1alpha1
+kind: GenerateUpgradePlanResponse
+status: Success # or Failure
+message: "error message if status == Failure"
+controlPlaneVersions:
+- v1.30.0
+- v1.31.0
+- v1.32.3
+- v1.33.0
+ ```
+
+Note: in this case the system will infer the list of intermediate version for workers from the list of control plane versions, taking
+care of performing the minimum number of workers upgrade by taking into account the [Kubernetes version skew policy](https://kubernetes.io/releases/version-skew-policy/).
+
+Implementers of this runtime extension can also support more sophisticated use cases, e.g.
+
+- Go through more patch release for a minor if necessary, e.g., v1.30.0 -> v1.30.1 -> etc.
+
+  ```yaml
+  ...
+  controlPlaneVersions:
+  - v1.30.0
+  - v1.30.1
+  - ...
+  ```
+  
+Note: in this case the system will infer the list of intermediate version for workers from the list of control plane versions, taking
+care of performing the minimum number of workers upgrade by taking into account the [Kubernetes version skew policy](https://kubernetes.io/releases/version-skew-policy/).
+
+- Force workers to upgrade to specific versions, e.g., force workers upgrade to v1.30.0 when doing v1.29.0 -> v1.32.3
+  (in this example, worker upgrade to 1.30.0 is not required by the [Kubernetes version skew policy](https://kubernetes.io/releases/version-skew-policy/), so it would
+  be skipped under normal circumstances).
+
+  ```yaml
+  ...
+  controlPlaneVersions:
+  - v1.30.0
+  - v1.31.0
+  - v1.32.3
+  workersVersions:
+  - v1.30.0
+  ```
+  
+Note: in this case the system will take into consideration the provided `workersVersions`, but if required by the [Kubernetes version skew policy](https://kubernetes.io/releases/version-skew-policy/),
+also add necessary intermediate version for workers inferred from the list of control plane versions.
```

Comment:
I'm not sure if we should do that. I would assume that if someone specifies the list explicitly they want the upgrade to be done exactly that way.

This is relevant if folks have tested that for specific cases n-4 => n works and they want to avoid unnecessary additional worker upgrades